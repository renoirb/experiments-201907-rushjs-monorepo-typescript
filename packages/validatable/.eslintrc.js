const base = require('@renoirb/conventions-use-eslint')

const main = {
  ...base,
  rules: {
    ...base.rules,
    // TODO: Make this list smaller, not bigger
    // To try it out, comment one line, then run `rushx lint`.
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/no-use-before-define': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
  },
}

module.exports = main
