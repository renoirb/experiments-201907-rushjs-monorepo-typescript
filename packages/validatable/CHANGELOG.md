# Change Log - @renoirb/validatable

This log was last generated on Tue, 04 Aug 2020 00:09:41 GMT and should not be manually modified.

## 1.1.3
Tue, 04 Aug 2020 00:09:41 GMT

*Version update only*

## 1.1.2
Mon, 03 Aug 2020 23:53:08 GMT

*Version update only*

## 1.1.1
Tue, 14 Jul 2020 03:23:31 GMT

### Patches

- Upgrading deps

## 1.1.0
Mon, 29 Jun 2020 02:44:25 GMT

### Minor changes

- Upgrading deps, minor refactor

## 1.0.2
Thu, 02 Jan 2020 23:25:02 GMT

### Patches

- Make TypeScript and TSLib peerDependency

