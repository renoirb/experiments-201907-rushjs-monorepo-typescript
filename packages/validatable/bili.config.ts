import useBili from '@renoirb/conventions-use-bili'

const main = useBili(
  {
    input: 'src/index.js',
    bundleNodeModules: true,
    externals: [],
  },
  { firstYear: 2003 },
)

export default main
