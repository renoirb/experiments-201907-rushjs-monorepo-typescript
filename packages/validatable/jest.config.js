const base = require('@renoirb/conventions-use-jest')

const main = {
  ...base,
  prettierPath:
    '<rootDir>/node_modules/@renoirb/conventions-code-formatter/node_modules/prettier/bin-prettier.js',
}

module.exports = main
