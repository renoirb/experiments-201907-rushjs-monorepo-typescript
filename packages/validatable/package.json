{
  "name": "@renoirb/validatable",
  "version": "1.1.2",
  "description": "Add validation rules to an object, and ability to check if a given string passes the rules",
  "homepage": "https://github.com/renoirb/experiments-201908-rush-typescript-just-bili-monorepo/tree/v1.x-dev/tools/bundling-helpers",
  "repository": {
    "type": "git",
    "url": "https://github.com/renoirb/experiments-201908-rush-typescript-just-bili-monorepo.git",
    "directory": "packages/validatable"
  },
  "license": "MIT",
  "author": {
    "name": "Renoir Boulanger",
    "email": "contribs@renoirboulanger.com"
  },
  "exports": {
    "browser": "./dist/index-browser.js",
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "main": "dist/index.js",
  "jsdelivr": "dist/index-browser.js",
  "module": "dist/index.mjs",
  "browser": {
    "./dist/index.js": "./dist/index-browser.js"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "use-cross-env use-run-all clean build:*",
    "build:browser": "use-cross-env BROWSERSLIST='> 0.5%, last 2 versions' use-bili --config bili.config.ts --target browser --format iife --module-name Validatable --file-name index-browser.js",
    "build:node": "use-cross-env BROWSERSLIST='maintained node versions' use-bili --config bili.config.ts --target node --format cjs --file-name index.js",
    "build:node-esm": "use-cross-env BROWSERSLIST='maintained node versions' use-bili --config bili.config.ts --target node --format esm --file-name index.mjs",
    "clean": "use-cross-env use-rimraf .rpt2_cache dist *.*.log",
    "fix": "use-cross-env conventions-code-formatter prettier '**/*.{ts,md,json,js}' --write",
    "lint": "use-cross-env conventions-use-eslint --fix 'src/**/*.ts'",
    "prepublishOnly": "rushx build",
    "test": "use-cross-env conventions-use-jest --detectOpenHandles",
    "test:snapshots": "use-cross-env conventions-use-jest --detectOpenHandles -u"
  },
  "devDependencies": {
    "@babel/core": "^7.11.0",
    "@babel/plugin-transform-runtime": "^7.11.0",
    "@babel/preset-env": "^7.11.0",
    "@babel/preset-typescript": "^7.10.4",
    "@babel/runtime-corejs3": "^7.11.0",
    "@renoirb/conventions-code-formatter": "workspace:*",
    "@renoirb/conventions-use-bili": "workspace:*",
    "@renoirb/conventions-use-eslint": "workspace:*",
    "@renoirb/conventions-use-jest": "workspace:*",
    "@renoirb/conventions-use-typescript-3": "workspace:*",
    "@renoirb/tools-bundling-helpers": "workspace:*",
    "@types/jest": "^26.0.0",
    "@types/node": "^12.12.0",
    "core-js": "^3.6.0",
    "jest": "^26.2.0",
    "jest-date-mock": "^1.0.8",
    "ts-jest": "^26.2.0",
    "tslib": "^2.0.0",
    "typescript": "~3.9.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
