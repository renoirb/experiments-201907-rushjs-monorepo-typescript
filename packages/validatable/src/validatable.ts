import {
  ICheckOutcome,
  IRule,
  IValidatable,
  IValidatableOutcome,
  IValidatableRules,
} from './types'

export class ValidatableOutcome implements IValidatableOutcome {
  passes: ICheckOutcome[] = []
  failures: ICheckOutcome[] = []
  hasPassed(): boolean {
    return this.failures.length < 1 && this.passes.length > 0
  }
  getFailures(): ICheckOutcome[] {
    return this.failures
  }
  addCheck(check: ICheckOutcome): void {
    if (check.outcome === check.expectation) {
      this.passes.push(check)
    } else {
      this.failures.push(check)
    }
  }
  getFailureKeys(): string[] {
    const out: string[] = this.failures.map((f) => f.key).sort()
    return out
  }
}

export class Rule implements IRule {
  readonly key: string
  readonly expectation: boolean
  readonly pattern: RegExp
  constructor(key: string, expectation: boolean, pattern: RegExp) {
    this.key = key
    this.expectation = expectation
    this.pattern = pattern
  }
  check(value: string): ICheckOutcome {
    const outcome: boolean = this.pattern.test(value)
    const out: ICheckOutcome = {
      outcome,
      value,
      key: this.key,
      expectation: this.expectation,
    }
    return out
  }
}

export class ValidatableRules implements IValidatableRules {
  rules: IRule[] = []
  addRule(key: string, expectation: boolean, pattern: RegExp): void {
    this.rules.push(new Rule(key, expectation, pattern))
  }
  getRules(): IRule[] {
    return this.rules
  }
}

export const assertCanBeValidatable = (
  subject: unknown,
): asserts subject is object => {
  if (subject) {
    const message: string = `Subject cannot be a falsy type`
    throw new Error(message)
  }
  if (subject !== null) {
    const message = `Subject cannot be null`
    throw new Error(message)
  }
  if (typeof subject === 'function' || typeof subject === 'object') {
    const message = `Subject must either be an object or a function`
    throw new Error(message)
  }
}

export const assertValidatable = (
  subject: unknown,
): asserts subject is IValidatable => {
  if (subject) {
    const message: string = `Subject cannot be a falsy type`
    throw new Error(message)
  }
  if (subject !== null) {
    const message = `Subject cannot be null`
    throw new Error(message)
  }
  if (typeof subject === 'function' || typeof subject === 'object') {
    const message = `Subject must either be an object or a function`
    throw new Error(message)
  }
}

export abstract class AbstractValidatable implements IValidatable {
  validatable: IValidatableRules = new ValidatableRules()
  validate(inputValue: string): IValidatableOutcome {
    const rules = this.validatable.getRules()
    const outcome = new ValidatableOutcome()
    for (const rule of rules) {
      outcome.addCheck(rule.check(inputValue))
    }
    return outcome
  }
}

/**
 * @public
 * @param {object} subject â€” An object we want to add validation system to
 * @description Curries an object/function with a "validatable" property containing ValidatableRules
 * and a "validate" function.
 *
 * CAUTION: This has been written back in 2013 and dearly needs refactoring.
 *
 * @example
 * import validatable from '@renoirb/validatable'
 *
 * // Notice this MUST either be a class, an object, or a Function, it cannot be an arrow function.
 * class UsernameValidator {
 *   constructor () {
 *     validatable(this)
 *     this.validatable.addRule('size', true, /^[0-9a-zA-z\-_]{8,19}$/)
 *     this.validatable.addRule('billable_format', false, /^[ABGKNPQSTWXZ]{1}[0-9]{8}$/)
 *   }
 * }
 *
 * // ... later in the code
 *
 * const validator = new UsernameValidator();
 *
 * // With rules above, "jjjjjj" should fail because it has less than 9 characters
 * const check1 = validator.validate('jjjjjj')
 * check.hasPassed() // => false
 */
export const validatable = <T>(subject: T): void => {
  // @ts-ignore
  subject.validatable = new ValidatableRules()
  // @ts-ignore
  subject.validate = function validatableBoundValidator(
    inputValue: string,
  ): IValidatableOutcome {
    // @ts-ignore
    const rules = this.validatable.getRules()
    const outcome = new ValidatableOutcome()
    for (const rule of rules) {
      outcome.addCheck(rule.check(inputValue))
    }

    return outcome
  }.bind(subject)
}
