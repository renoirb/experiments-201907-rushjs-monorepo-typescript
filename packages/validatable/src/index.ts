export * from './validatable'
export * from './password'
export * from './types'

import { validatable } from './validatable'

export default validatable
