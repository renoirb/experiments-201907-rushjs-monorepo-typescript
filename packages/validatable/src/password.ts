import { validatable, AbstractValidatable } from './validatable'

export class PasswordValidator extends AbstractValidatable {
  constructor() {
    super()
    validatable(this)
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes
    this.validatable.addRule('mustBeLength', true, /^.{8,19}$/)
    this.validatable.addRule('mustContainCapitals', true, /[A-Z]/)
    this.validatable.addRule('mustContainDigits', true, /\d{1,}/)
    this.validatable.addRule(
      'mustContainSpecial',
      true,
      /[\u00C0-\u017F\s\u00C0-\u017F!@#$%?&*\(\)\[\]\{\}\-\+=:~]{1,}/,
    )
    this.validatable.addRule('mustNotRepeatCharacters', false, /(.)\1{2,}/)
  }
}
