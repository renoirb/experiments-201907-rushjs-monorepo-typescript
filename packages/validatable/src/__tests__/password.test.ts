// @ts-nocheck

import { PasswordValidator } from '../password'
import { ValidatableRules } from '../validatable'

describe('Sanity checks', () => {
  test('Subject that has been curried has validatable object', () => {
    const subject = new PasswordValidator()
    expect(subject).toMatchSnapshot()
    expect(subject).toHaveProperty('validatable')
    expect(subject.validatable).toBeInstanceOf(ValidatableRules)
    expect(subject).toHaveProperty('validate')
  })
})

describe('PasswordValidator rules dry-run', () => {
  const subject = new PasswordValidator()
  const runs = []

  const ek = [
    /* 0 */ 'mustBeLength',
    /* 1 */ 'mustContainCapitals',
    /* 2 */ 'mustContainDigits',
    /* 3 */ 'mustContainSpecial',
    /* 4 */ 'mustNotRepeatCharacters',
  ]

  runs.push(['jjjjjjj', false, 'case 1: Less than minimum length', [...ek]])
  runs.push([
    'jjjjjjjj',
    false,
    'case 2: Only minimum length',
    [ek[1], ek[2], ek[3], ek[4]],
  ])
  runs.push([
    'j1jjjjjj',
    false,
    'case 3: Minimum length, one digit, yet insufficient',
    [ek[1], ek[3], ek[4]],
  ])
  runs.push([
    'j1jj%jjj',
    false,
    'case 4: Minimum length, one digit, one special, yet insufficient',
    [ek[1], ek[4]],
  ])
  runs.push(['j1Jj%jjj', false, 'case 5: Repetition', [ek[4]]])
  runs.push([
    'j1Jj!!!-',
    false,
    'case 5: Repetition even when special',
    [ek[4]],
  ])
  runs.push(['j1Jj%jjT', true, 'case 6: All requirements met', []])

  runs.push(['@CSIs33lf@!!', true, 'case 6: All requirements edge-case', []])

  // Known edge-case, unicode characters might pass if repeated
  // runs.push(['j1JjðŸ˜ºðŸ˜ºðŸ˜º-', false, 'case 7: Repetition even when special', [ek[4]]])

  for (const [
    password,
    expectedBoolean,
    exampleSummary,
    expectedFailureKeys,
  ] of runs) {
    const check = subject.validate(password)
    const failures = check.getFailureKeys().sort()
    const expectedFailureCodes = expectedFailureKeys.sort()
    const expectedLength = expectedFailureCodes.length
    const expectedCodesStr = expectedFailureCodes.join(',')
    const testExpectedErrorCodes = `${exampleSummary}: "${password}",  expected ${expectedLength} failures, error codes should be [${expectedCodesStr}], but we got [${failures.join(
      ',',
    )}] }`
    it(testExpectedErrorCodes, () => {
      expect(failures).toEqual(expect.arrayContaining(expectedFailureCodes))
      expect(failures).toHaveLength(expectedLength)
    })
    const testSummaryExplanation = `${exampleSummary}: "${password}",  should be ${expectedBoolean.toString()}`
    it(testSummaryExplanation, () => {
      expect(check.hasPassed()).toBe(expectedBoolean)
    })
  }
})
