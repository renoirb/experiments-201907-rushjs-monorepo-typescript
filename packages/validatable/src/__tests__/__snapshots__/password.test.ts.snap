// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity checks Subject that has been curried has validatable object 1`] = `
PasswordValidator {
  "validatable": ValidatableRules {
    "rules": Array [
      Rule {
        "expectation": true,
        "key": "mustBeLength",
        "pattern": /\\^\\.\\{8,19\\}\\$/,
      },
      Rule {
        "expectation": true,
        "key": "mustContainCapitals",
        "pattern": /\\[A-Z\\]/,
      },
      Rule {
        "expectation": true,
        "key": "mustContainDigits",
        "pattern": /\\\\d\\{1,\\}/,
      },
      Rule {
        "expectation": true,
        "key": "mustContainSpecial",
        "pattern": /\\[\\\\u00C0-\\\\u017F\\\\s\\\\u00C0-\\\\u017F!@#\\$%\\?&\\*\\\\\\(\\\\\\)\\\\\\[\\\\\\]\\\\\\{\\\\\\}\\\\-\\\\\\+=:~\\]\\{1,\\}/,
      },
      Rule {
        "expectation": false,
        "key": "mustNotRepeatCharacters",
        "pattern": /\\(\\.\\)\\\\1\\{2,\\}/,
      },
    ],
  },
  "validate": [Function],
}
`;
